### PR TITLE
Bug #4316: fix proposal for undefined references when BUILD_opencv_world:BOOL=ON

### PR DESCRIPTION
**Pull request by lupustr3**
I don't know if you specifically don't want to export this functions. But I don't see any reason not to do this since they are included in API headers anyway.

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
> I don't know if you specifically don't want to export this functions. But I don't see any reason not to do this since they are included in API headers anyway.

I believe these functions are internal and targeted for usage inside of public exported OpenCV functions only. E.g. `hal::exp` should be used inside `cv::exp` and so on.

File: modules/hal/include/opencv2/hal/defs.h [hunk 1]
```diff
@@ -45,6 +45,8 @@
 #ifndef __OPENCV_DEF_H__
 #define __OPENCV_DEF_H__
 
+#include "opencv2/core/cvdef.h"
```
Comment (ilya-lavrenov) [inline modules/hal/include/opencv2/hal/defs.h]: `HAL` is a self-contained module and must not depend on others

**Comment by lupustr3** [comment]
Shouldn't then hal.hpp be included in private.hpp instead of base.hpp? It doesn't make sense to expose declaration but prevent definition export.

**Comment by lupustr3** [comment]
So... I want to find solution which can be adopted by Itseez. Do you think that excluding hal module from world and building it separately can be suitable? Then it will behave like non-world build (link statically).

**Comment by alalek** [comment]
I'm not sure that exports of HAL functions is a part of design. HAL was designed as a static library initially without strict restrictions to ABI (and may be source level).
I believe @vpisarev or @mshabunin can describe/clear these moments more correctly.

**Pull request closed without merging**